File: tms/turacoz_management_system/report/freelancer_alignment_report/freelancer_alignment_report.py
# Copyright (c) 2022, RSA and contributors
# For license information, please see license.txt

# import frappe

from __future__ import unicode_literals
from frappe import _
import frappe
from datetime import datetime
from datetime import timedelta
import calendar
import pendulum
import datetime



def execute(filters=None):
	columns, data = [], []
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_data(filters):
	freelancer = filters.get("freelancer")
	task_status = filters.get("task_status")
	data = []
	
	if freelancer:
		freelancer1 = " and tf.name = '%s'" %freelancer
	else:
		freelancer1 = ''	
	
	data1 = frappe.db.sql("""select tf.name as freelancer_id,tf.freelancer_name,tfpd.freelancer_feedback_project_id as freelancer_proj_id,tfpd.project,tfpd.role,tfpd.type_of_documents,
			tfpd.start_date as task_start_date,tfpd.project_close_date as task_end_date,tp.project_title,tp.customer,
			tp.project_current_status,tp.expected_end_date,tfpd.status,tfpd.actual_close_date,tp.manager_name from `tabFreelancer` tf 
			left join `tabFreelancer Feedback Project Detail` tfpd on tf.name =tfpd.parent
			left join `tabProject` tp on tfpd.project = tp.name 
			where tp.project_current_status not in ('Completed','Completed Technically','Cancelled') and tfpd.status='{0}' {1};""".format(task_status,freelancer1), as_dict=True)
	
	for rec in data1:
		rows = {}
		rows['name'] = rec.freelancer_id
		rows['freelancer_name'] = rec.freelancer_name
		rows['project'] = rec.project
		rows['manager_name'] = rec.manager_name
		rows['task_status'] = rec.status
		rows['role'] = rec.role
		rows['deliverable'] = rec.type_of_documents
		rows['task_start_date'] = rec.task_start_date
		rows['task_end_date'] = rec.task_end_date
		rows['project_title'] = rec.project_title
		rows['client'] = rec.customer
		rows['project_status'] = rec.project_current_status
		rows['expected_end_date'] = rec.expected_end_date
		rows['action'] = '<button style=''color:white;background-color:green;'' type=''button'' onClick= ''consoleerp_hi("{0}","{1}","{2}","{3}","{4}","{5}")''>Action</button>' .format(rec.freelancer_proj_id,rec.project,rec.task_start_date,rec.task_end_date,rec.status,rec.actual_close_date)
		
		data.append(rows)
		
	return data	
	
def get_columns(filters):
	freelancer = filters.get("freelancer")
	cols = []
	cols=[
			{
				"fieldname": "name",
				"label": ("Freelancer ID"),
				"fieldtype": "Link",
				"options": "Freelancer",
				"width": "150"
			},
			{
				"fieldname": "freelancer_name",
				"label": ("Freelancer Name"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "project",
				"label": ("Project"),
				"fieldtype": "Link",
				"options": "Project",
				"width": "150"
			},
			{
				"fieldname": "task_status",
				"label": ("Task Status"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "role",
				"label": ("Role"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "deliverable",
				"label": ("Deliverable"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "task_start_date",
				"label": ("Task Start Date"),
				"fieldtype": "Date",
				"width": "150"
			},
			{
				"fieldname": "task_end_date",
				"label": ("Task End Date"),
				"fieldtype": "Date",
				"width": "150"
			},
			{
				"fieldname": "manager_name",
				"label": ("Project Manager"),
				"fieldtype": "Data",
				"width": "150"
			},			
			{
				"fieldname": "project_title",
				"label": ("Project Title"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "client",
				"label": ("Client"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "project_status",
				"label": ("Project Status"),
				"fieldtype": "Data",
				"width": "150"
			},
			{
				"fieldname": "expected_end_date",
				"label": ("Project Expected End Date"),
				"fieldtype": "Date",
				"width": "150"
			},
			{
				"fieldname": "action",
				"label": ("Action"),
				"fieldtype": "Button",
				"width": "150"
			}
		]
	return cols	

@frappe.whitelist()
def update_planned(proje_id,status,actual_close_date):

 	data1 = frappe.db.sql("""update `tabFreelancer Feedback Project Detail` set status='{0}',actual_close_date='{1}' where freelancer_feedback_project_id={2};""".format(status,actual_close_date,proje_id))
 	
 	return 1

File: tms/turacoz_management_system/report/freelancer_alignment_report/freelancer_alignment_report.js
// Copyright (c) 2022, RSA and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Freelancer Alignment Report"] = {
	"filters": [
	
		{
			"fieldname": "freelancer",
			"label": __("Freelancer"),
			"fieldtype": "Link",
			"options": "Freelancer"
		},
		{
			"fieldname": "task_status",
			"label": __("Task Status"),
			"fieldtype": "Select",
			"options": ['Open','Close','Cancelled','All'],
			"default": "Open",
			"reqd": 1
		},


	]
};

function consoleerp_hi(freelance_project_table_id,project,start_date,end_date,status,actual_close_date) {
	
	var d = new frappe.ui.Dialog({
	    'fields': [
		{'fieldname': 'proje_id', 'fieldtype': 'Int', 'placeholder': 'Project Table ID'},
		{'fieldname': 'project', 'fieldtype': 'Data', 'label': 'Project','placeholder': 'Project'},
		{'fieldname': 'status', 'fieldtype': 'Select', 'placeholder': 'Status', 'label': 'Task Status', 'options':['Open','Close','Cancelled']},
		{'fieldname': 'start_date', 'fieldtype': 'Date', 'placeholder': 'Start Date', 'label': 'Align Start Date'},
		{'fieldname': 'end_date', 'fieldtype': 'Date', 'placeholder': 'End Date', 'label': 'Align End Date'},
		{'fieldname': 'actual_close_date', 'fieldtype': 'Date', 'placeholder': 'Actual Close Date', 'label': 'Actual Close Date'}
	    ],
	    primary_action: function(frm){
		d.hide();
		//show_alert(d.get_values());
		
	    }
	});
	if(start_date){
		s_date = start_date;
	}
	else{
		s_date = convert.ToDateTime('01-01-1900');
	}
	
	if(end_date){
		e_date = end_date;
	}
	else{
		e_date = '1900-01-01';
	}
	if(actual_close_date){
		ac_date = actual_close_date;
	}
	else{
		ac_date = '1900-01-01';
	}
	d.fields_dict.proje_id.set_value(freelance_project_table_id);
	d.fields_dict.project.set_value(project);	
	d.fields_dict.status.set_value(status);
	d.fields_dict.start_date.set_value(s_date);
	d.fields_dict.end_date.set_value(e_date);	
	d.fields_dict.actual_close_date.set_value(ac_date);	
	d.show();
	
	d.set_primary_action(__("Submit"), function() {
		var values = d.get_values();
		if(values.actual_close_date){
			var actual_close_date1 = values.actual_close_date;
		}
		else{
			var actual_close_date1 = '';
		}

	
		frappe.call({
				method: "tms.turacoz_management_system.report.freelancer_alignment_report.freelancer_alignment_report.update_planned",
				args:{ "proje_id":values.proje_id,"status":values.status,"actual_close_date":actual_close_date1
			},
			callback:function(res){
				let a = res.message; 
				if (a!=0){
					show_alert("Task Status Successfully Updated")
				}
				else{
					show_alert("Plan inserted for this task")
				}
			}
		})
	})
}	
